collections/src/main/scala/strawman/collection/Iterable.scala

```diff
@@ -878,7 +877,30 @@ trait IterableOps[+A, +CC[X], +C] extends Any {
     val unzipped = View.Unzip(toIterable)
     (fromIterable(unzipped.first), fromIterable(unzipped.second))
   }
-
 }
 
-object Iterable extends IterableFactory.Delegate[Iterable](immutable.Iterable)
+object Iterable extends IterableFactory.Delegate[Iterable](immutable.Iterable) {
+
+  implicit class LazyZipOps[A, C1[X] <: Iterable[X]](`this`: C1[A]) {
+
+    /** Analogous to `zip` except that the elements in each collection are not consumed until a strict operation is
+      * invoked on the returned `Tuple2Zipped` decorator.
+      *
+      * Calls to `lazyZip` can be chained to support higher arities (up to 4) without incurring the expense of
+      * constructing and deconstructing intermediary tuples.
+      *
+      *    {{{
+      *    val xs = $Coll(1, 2, 3)
+      *    val res = (xs lazyZip xs lazyZip xs lazyZip xs).map((a, b, c, d) => a + b + c + d)
+      *    // res == $Col(4, 8, 12)
+      *    }}}
+      *
+      * @param that the iterable providing the second half of each eventual pair
+      * @tparam B the type of the element in the second half of each eventual pair
+      * @tparam C2 the type of `that` iterable
+      * @return a decorator `Tuple2Zipped` that allows strict operations to be performed on the lazily evaluated pairs
+      *         or chained calls to `lazyZip`. Implicit conversion to `Iterable[(A, B)]` is also supported.
+      */
+    def lazyZip[B, C2[X] <: Iterable[X]](that: C2[B]): Tuple2Zipped[A, C1[A], B, C2[B]] = new Tuple2Zipped(`this`, that)
```

Comment:
By convention we use the `CC` identifier for **C**ollection type **C**onstructors, while `C` is used for **C**ollection types. BTW, I’m wondering, here, what’s the advantage of using a collection type constructor over a collection type?

Reply:
>I’m wondering, here, what’s the advantage of using a collection type constructor over a collection type?

```
scala> import strawman.collection._
import strawman.collection._

scala> import strawman.collection.immutable._
import strawman.collection.immutable._

scala> val xs: List[Int] = List(1, 2, 3)
xs: strawman.collection.immutable.List[Int] = List(1, 2, 3)

scala> xs.lazyZip(xs)
<console>:19: error: value lazyZip is not a member of strawman.collection.immutable.List[Int]
       xs.lazyZip(xs)
          ^
```

Using collection type in `TupleNZipped.lazyZip`:

```
scala> xs.lazyZip(xs).lazyZip(xs)
<console>:19: error: inferred type arguments [Nothing,strawman.collection.immutable.List[Int]] do not conform to method lazyZip's type parameter bounds [B,C3 <: strawman.collection.Iterable[B]]
       xs.lazyZip(xs).lazyZip(xs)
                      ^
<console>:19: error: type mismatch;
 found   : strawman.collection.immutable.List[Int]
 required: C3
       xs.lazyZip(xs).lazyZip(xs)
                              ^
```

I got confused with this one as well. I started with the signature suggested in one of your last comments, but then I faced these errors. This was pretty much the reason I used implicit conversions in the first iteration. Am I missing anything?

Reply:
Hmm ok I see. Thanks for the explanation!